rpc/types.go

```diff
@@ -143,45 +138,41 @@ const (
 // - an out of range error when the given block number is either too little or too large
 func (bn *BlockNumber) UnmarshalJSON(data []byte) error {
 	input := strings.TrimSpace(string(data))
-
 	if len(input) >= 2 && input[0] == '"' && input[len(input)-1] == '"' {
 		input = input[1 : len(input)-1]
 	}
 
-	if len(input) == 0 {
-		*bn = BlockNumber(latestBlockNumber.Int64())
+	switch input {
+	case "earliest":
+		*bn = EarliestBlockNumber
+		return nil
+	case "latest":
+		*bn = LatestBlockNumber
+		return nil
+	case "pending":
+		*bn = PendingBlockNumber
 		return nil
 	}
 
-	in := new(big.Int)
-	_, ok := in.SetString(input, 0)
-
-	if !ok { // test if user supplied string tag
-		strBlockNumber := input
-		if strBlockNumber == "latest" {
-			*bn = BlockNumber(latestBlockNumber.Int64())
-			return nil
-		}
-
-		if strBlockNumber == "earliest" {
-			*bn = BlockNumber(earliestBlockNumber.Int64())
-			return nil
-		}
+	if len(input) < 3 {
+		return fmt.Errorf("invalid blocknumber %s", data)
+	}
+	if input[0] != '0' || (input[1] != 'x' && input[1] != 'X') {
+		return fmt.Errorf("blocknumber must start with 0x")
+	}
+	if input[2] == '0' && len(input) != 3 {
+		return fmt.Errorf("invalid blocknumber %s", data)
+	}
```

Comment:
Okay, I rewrote it so it uses the hexutil package. There is just one thing that I'm not sure about. The json spec afaik doesn't say anything about how an block number should be formatted in case it is a number, it just describes it as a hex string. Usually the term `QUANTITY` is used, in this case it is defined as a hex string. That would mean that leading zero's should be allowed. I checked and parity seem to allow leading zero's.